Add Parser, ModHappyParser, ExitCommand

Ch40gRv1-Mu opened the pull request:
Parser is the parent class of all parser, that provide method to parse any format of string; ModHappyParser is the parser to parse general format of command;

Ch40gRv1-Mu committed:
Add Parser, ModHappyParser, ExitCommand

Parser is the parent class of all parser, that provide method to parse any format of string; ModHappyParser is the parser to parse general format of command;

chooyikai commented on src/main/java/seedu/duke/exceptions/UnkownCommandException.java:
```diff
@@ -0,0 +1,13 @@
+package seedu.duke.exceptions;
+
+public class UnkownCommandException extends ModHappyException {
+    private static final String ERROR_MESSAGE = "Sorry, I don't understand the following command(,,•́ . •̀,,) :";
```
I tested this in the command line and `•́ . •̀ `does not print properly... maybe change it to something else? (Same for the other places this face is used)

Ch40gRv1-Mu replied:
Hi, may I check how the face look in your side, because it works from my side... How about I just remove all the faces first?

chooyikai replied:
Let's remove all the faces first and decide later... the face on my side appears like `•?. •̀ `

chooyikai commented on src/main/java/seedu/duke/exceptions/UnkownCommandException.java:
```diff
@@ -0,0 +1,13 @@
+package seedu.duke.exceptions;
+
+public class UnkownCommandException extends ModHappyException {
```
`UnkownCommandException` should be `UnknownCommandException`

chooyikai commented on src/main/java/seedu/duke/parsers/ModHappyParser.java:
```diff
@@ -0,0 +1,47 @@
+    private static final String ARGUMENT = "arguments";
+    private static final String COMMAND_WORD = "commandWord";
+    private static final String EXIT_COMMAND_WORD = "exit";
+    private static final String MOD_HAPPY_COMMAND_FORMAT = "(?<commandWord>\\S+)(?<arguments>.*)";
```
The current regex will parse "hello argument" into "hello" and " argument" (note the space before "argument"). Maybe change this to `(?<commandWord>\\S+)( (?<arguments>.*))` instead?

chooyikai commented on src/main/java/seedu/duke/parsers/ModHappyParser.java:
```diff
@@ -0,0 +1,47 @@
+    private static final String MOD_HAPPY_COMMAND_FORMAT = "(?<commandWord>\\S+)(?<arguments>.*)";
+
+
+    public ModHappyParser() throws ModHappyException {
```
Not sure why this constructor `throws ModHappyException` because I could not find any code inside it which does that. Can we remove it?

chooyikai commented on src/main/java/seedu/duke/commands/Command.java:
```diff
@@ -1,14 +1,17 @@
 public abstract class Command {
     protected String commandName = "Command";
 
-    public String execute() throws InstantiateAbstractClassException {
```
Since `InstantiateAbstractClassException` is not used anymore (and anyway Java has its own built in exception for this) maybe we can just delete the class

chooyikai reviewed (changes requested):
Other general comments:
- Let's add a comment on the parts of the code that were reused from AddressBook (the module webpage says we must acknowledge)
- Don't forget to write JUnit tests (although I am not sure if there is enough functional code to write anything meaningful yet, maybe 1 or 2 lines is ok)

Ch40gRv1-Mu committed:
Fix the regex in ModHappyParser

Change from (?<commandWord>\\S+)(?<arguments>.*) to (?<commandWord>\\S+)\\s*(?<arguments>.*) to fix the possible space between commanWord and arguments

Ch40gRv1-Mu committed:
Remove unnecessary exception in ModHappyParser

Remove unnecessary exception in ModHappyParser

Ch40gRv1-Mu committed:
Add reference for runCommandLoopUntilExitCommand and exit in Main

Add the link to reference in the Java Doc for runCommandLoopUntilExitCommand and exit in Main

Ch40gRv1-Mu committed:
Refactor the name for UnknownCommandException

Refactor name for UnknownCommandException

chooyikai commented:
Okay, LGTM
